docs/rfcs/036-bottom-most-gc-compaction.md

```diff
@@ -0,0 +1,182 @@
+
+![](images/036-bottom-most-gc-compaction/08-optimization.svg)
+
+Consider the case that the system finishes bottom-most compaction at GC horizon 0x60. Now GC horizon grows to 0x70 and the branch at 0x50 is deleted. We only need to pick the layers between 0x40 and 0x70 for compaction (the red layers). As other lower layers will not change after the compaction process, it does not need to be picked.
+
+# Result
+
+Bottom-most compaction ensures all garbages under the GC horizon gets collected right away (compared with “eventually” in the current algorithm). Meanwhile, it generates images at each of the retain_lsn to ensure branch reads are fast. As we make per-key decision on whether to generate an image or not, the theoretical lower bound of the storage space we need to retain a branch is lower than before.
+
+Before: min(sum(logs for each key), sum(image for each key)), for each partition — we always generate image layers on a key range
+
+After: sum(min(logs for each key, image for each key))
+
+# Dealing with Large Input
+
+Bottom-most compaction does a full compaction below the GC horizon, and the process is currently designed as non-resumable. The process should take a fixed amount of memory and be able to run for a few minutes.
+
+We made two design choices for the compaction algorithm:
+
+- Use k-merge to collect data required for the compaction. The memory consumption is `<buffered key-values per file> x <num of layers involved in compaction>`.
+    - Key history of 1 million → OOM?
+- Parallel compaction to make full use of the CPU resources.
+
+The image creation is integrated into the k-merge, i.e., reconstruct data naturally becomes available as part of k-merge instead of calling into get_values_reconstruct_data. This is significantly more CPU efficient than the image layer creation of legacy compaction.
+
+# Compaction Trigger
+
+The bottom-most compaction should be automatically triggered. The goal of the trigger is that it should ensure a constant factor for write amplification. Say that the user write 1GB of WAL into the system, we should write 1GB x C data to S3. The legacy compaction algorithm does not has such a constant factor C. The data we write to S3 is quadratic to the logical size of the database (see [A Theoretical View of Neon Storage](https://www.notion.so/A-Theoretical-View-of-Neon-Storage-8d7ad7555b0c41b2a3597fa780911194?pvs=21)).
+
+We propose the following compaction trigger that generates a constant write amplification factor. Write amplification = total writes to S3 / total user writes. We only analyze the write amplification caused by the bottom-most GC-compaction process, ignoring the legacy create image layers amplification.
```

Comment:
```suggestion
We propose the following compaction trigger that generates a constant write amplification factor. Write amplification >= total writes to S3 / total user writes. We only analyze the write amplification caused by the bottom-most GC-compaction process, ignoring the legacy create image layers amplification.
```

Reply:
Also, what is "total user writes" in this context? The physical size? The amount of WAL? So basically the lsn?

How do we determine/store "total writes to S3", because we will delete stuff from there regularly.

Or does "total" mean just the stuff inside the layer map?